File: worker/live/overlay.py
"""Overlay rendering for live sessions using supervision annotators."""
from __future__ import annotations

from dataclasses import dataclass, field

from vision.types import CLASS_BALL, CLASS_PLAYER, CLASS_REFEREE, FrameResult


@dataclass
class OverlayFlags:
    boxes: bool = True
    labels: bool = True
    traces: bool = False

    def any_overlay(self) -> bool:
        return self.boxes or self.labels or self.traces


_COLOR_MAP = {
    CLASS_PLAYER: (255, 100, 30),    # orange-blue (BGR)
    CLASS_BALL: (0, 220, 255),       # yellow (BGR)
    CLASS_REFEREE: (0, 0, 220),      # red (BGR)
    "goalkeeper": (0, 200, 0),       # green (BGR)
}
_DEFAULT_COLOR = (200, 200, 200)


def draw_overlay(frame, fr: FrameResult, flags: OverlayFlags) -> None:
    """Draw bounding boxes and labels onto frame in-place (BGR numpy array)."""
    import cv2

    for det in fr.detections:
        x1, y1, x2, y2 = (int(v) for v in det.xyxy)
        color = _COLOR_MAP.get(det.cls, _DEFAULT_COLOR)

        if flags.boxes:
            cv2.rectangle(frame, (x1, y1), (x2, y2), color, 2)

        if flags.labels:
            label = det.cls
            if det.track_id is not None:
                label = f"{det.cls} #{det.track_id}"
            cv2.putText(
                frame,
                label,
                (x1, max(y1 - 6, 10)),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.45,
                color,
                1,
                cv2.LINE_AA,
            )

File: worker-dotnet/PitchWise.Vision/Overlay.cs
using OpenCvSharp;

namespace PitchWise.Vision;

/// <summary>Which overlay layers to draw on a live frame. Port of overlay.py OverlayFlags.</summary>
public sealed class OverlayFlags
{
    public bool Boxes { get; set; } = true;
    public bool Labels { get; set; } = true;
    public bool Traces { get; set; }

    public bool AnyOverlay() => Boxes || Labels || Traces;
}

/// <summary>Bounding-box + label overlay rendering. Port of worker/live/overlay.py.</summary>
public static class Overlay
{
    // BGR colors per class (matches the Python _COLOR_MAP).
    private static readonly Dictionary<string, Scalar> ColorMap = new()
    {
        [ObjectClass.Player] = new Scalar(255, 100, 30),   // orange-blue
        [ObjectClass.Ball] = new Scalar(0, 220, 255),      // yellow
        [ObjectClass.Referee] = new Scalar(0, 0, 220),     // red
        [ObjectClass.Goalkeeper] = new Scalar(0, 200, 0),  // green
    };
    private static readonly Scalar DefaultColor = new(200, 200, 200);

    /// <summary>Draws boxes and labels onto <paramref name="frame"/> in place (BGR Mat).</summary>
    public static void Draw(Mat frame, FrameResult fr, OverlayFlags flags)
    {
        foreach (Detection det in fr.Detections)
        {
            int x1 = (int)det.X1, y1 = (int)det.Y1, x2 = (int)det.X2, y2 = (int)det.Y2;
            Scalar color = ColorMap.TryGetValue(det.Cls, out Scalar c) ? c : DefaultColor;

            if (flags.Boxes)
                Cv2.Rectangle(frame, new Point(x1, y1), new Point(x2, y2), color, thickness: 2);

            if (flags.Labels)
            {
                string label = det.TrackId is int id ? $"{det.Cls} #{id}" : det.Cls;
                Cv2.PutText(
                    frame, label, new Point(x1, Math.Max(y1 - 6, 10)),
                    HersheyFonts.HersheySimplex, 0.45, color, thickness: 1,
                    lineType: LineTypes.AntiAlias);
            }
        }
    }
}
